### PR TITLE
Faster Pipeline Resolve

### PR DESCRIPTION
**Pull request by pmuetschard**
Instead of resolving the bound pipelines by getting all Pipeline resource datas and then filtering to the bound ones, use the state to find the bound pipelines. This avoids having to do another mutate run to resolve all the pipeline resource datas, most of which are then summarily ignored.

For a resource heavy trace, this brings the time it takes to get the bound pipelines on command selection in the UI from ~5 seconds to well below one second.

### CODE REVIEW

File: gapis/api/pipeline.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2020 Google Inc.
```
Comment (hevrard) [inline gapis/api/pipeline.go:1]: 2021

File: gapis/api/vulkan/draw_call_pipeline.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2020 Google Inc.
```
Comment (hevrard) [inline gapis/api/vulkan/draw_call_pipeline.go:1]: 2021

File: gapis/api/vulkan/draw_call_pipeline.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+	lastQueue := c.LastBoundQueue()
+	if lastQueue.IsNil() {
+		return bound, fmt.Errorf("No previous queue submission")
+	}
+
+	if cmd.CmdFlags().IsExecutedDraw() {
+		lastDrawInfo, ok := c.LastDrawInfos().Lookup(lastQueue.VulkanHandle())
+		if !ok {
+			return bound, fmt.Errorf("There have been no previous draws")
```
Comment (hevrard) [inline gapis/api/vulkan/draw_call_pipeline.go:60]: typo: no previous draws -> no previous draw

File: gapis/api/vulkan/draw_call_pipeline.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		}
+		bound.Pipeline = lastDrawInfo.GraphicsPipeline()
+	} else {
+		lastComputeInfo, ok := c.LastComputeInfos().Lookup(lastQueue.VulkanHandle())
+		if !ok {
+			return bound, fmt.Errorf("There have been no previous draws")
```
Comment (hevrard) [inline gapis/api/vulkan/draw_call_pipeline.go:66]: maybe rephrase: `no previous compute dispatch` ?

File: gapis/resolve/pipeline.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
+// Copyright (C) 2020 Google Inc.
```
Comment (hevrard) [inline gapis/resolve/pipeline.go:1]: 2021

File: gapis/resolve/pipeline.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	obj, err := ResolveInternal(ctx, p.Parent(), r)
+	if err != nil {
+		return nil, err
+	}
+
+	if pipelines, err := pipelinesFor(ctx, obj, p, r); err != api.ErrPipelineNotAvailable {
+		return pipelines, err
+	}
+	return nil, &service.ErrDataUnavailable{Reason: messages.ErrNotADrawCall()}
+}
```
Comment (hevrard) [inline gapis/resolve/pipeline.go:39]: Here and in other places, the api.ErrPipelineNotAvailable is replaced by messages.ErrNotADrawCall, I'm not sure why we don't simply forward the api.ErrPipelineNotAvailable error?

**Review by hevrard — APPROVED** [review]
LGTM with nits

File: gapis/resolve/pipeline.go (hunk 6 above)
Comment (pmuetschard) [reply gapis/resolve/pipeline.go:39]: `ErrPipelineNotAvailable` is not a "service error" and would be wrapped inside an `ErrInternalError`. However, we want to return the `ErrDataUnavailable` error here, because it has a specific semantic meaning to the client. `ErrPipelineNotAvailable` is not so much an "error" but a "null" indicator, kind of like `io.EOF`, i.e. not unexpected.

File: gapis/api/vulkan/draw_call_pipeline.go (hunk 3 above)
Comment (pmuetschard) [reply gapis/api/vulkan/draw_call_pipeline.go:60]: That would be grammatically incorrect ("There have been" is plural).

File: gapis/api/vulkan/draw_call_pipeline.go (hunk 4 above)
Comment (pmuetschard) [reply gapis/api/vulkan/draw_call_pipeline.go]: Done.

File: gapis/resolve/pipeline.go (hunk 5 above)
Comment (pmuetschard) [reply gapis/resolve/pipeline.go]: Done.

File: gapis/api/pipeline.go (hunk 1 above)
Comment (pmuetschard) [reply gapis/api/pipeline.go]: Done.

File: gapis/api/vulkan/draw_call_pipeline.go (hunk 2 above)
Comment (pmuetschard) [reply gapis/api/vulkan/draw_call_pipeline.go]: Done.
